src/audio_core/opus/hardware_opus.cpp

```diff
@@ -0,0 +1,227 @@
+    Service::Audio::ResultLibOpusUnimplemented, Service::Audio::ResultLibOpusInvalidPacket,
+    Service::Audio::ResultLibOpusInternalError, Service::Audio::ResultBufferTooSmall,
+    Service::Audio::ResultLibOpusBadArg,        ResultSuccess,
+};
+} // namespace
+
+HardwareOpus::HardwareOpus(Core::System& system_)
+    : system{system_}, opus_decoder{system.AudioCore().ADSP().OpusDecoder()} {
+    opus_decoder.SetSharedMemory(shared_memory);
+}
+
+u64 HardwareOpus::GetWorkBufferSize(s32 channel) {
+    if (!opus_decoder.IsRunning()) {
+        return 0;
+    }
+    std::scoped_lock l{mutex};
+    shared_memory.host_send_data[0] = (u32)channel;
+    opus_decoder.Send(ADSP::Direction::DSP, ADSP::OpusDecoder::Message::GetWorkBufferSize);
+    auto msg = opus_decoder.Receive(ADSP::Direction::Host);
+    if (msg != ADSP::OpusDecoder::Message::GetWorkBufferSizeOK) {
+        LOG_ERROR(Service_Audio, "OpusDecoder returned invalid message. Expected {} got {}",
+                  ADSP::OpusDecoder::Message::GetWorkBufferSizeOK, msg);
+        return 0;
+    }
+    return shared_memory.dsp_return_data[0];
+}
+
+u64 HardwareOpus::GetWorkBufferSizeForMultiStream(s32 total_stream_count, s32 stereo_stream_count) {
+    std::scoped_lock l{mutex};
+    shared_memory.host_send_data[0] = (u32)total_stream_count;
+    shared_memory.host_send_data[1] = (u32)stereo_stream_count;
+    opus_decoder.Send(ADSP::Direction::DSP,
+                      ADSP::OpusDecoder::Message::GetWorkBufferSizeForMultiStream);
+    auto msg = opus_decoder.Receive(ADSP::Direction::Host);
+    if (msg != ADSP::OpusDecoder::Message::GetWorkBufferSizeForMultiStreamOK) {
+        LOG_ERROR(Service_Audio, "OpusDecoder returned invalid message. Expected {} got {}",
+                  ADSP::OpusDecoder::Message::GetWorkBufferSizeForMultiStreamOK, msg);
+        return 0;
+    }
+    return shared_memory.dsp_return_data[0];
+}
+
+Result HardwareOpus::InitializeDecodeObject(s32 sample_rate, s32 channel_count, void* buffer,
+                                            u64 buffer_size) {
+    std::scoped_lock l{mutex};
+    shared_memory.host_send_data[0] = (u64)buffer;
+    shared_memory.host_send_data[1] = buffer_size;
+    shared_memory.host_send_data[2] = sample_rate;
+    shared_memory.host_send_data[3] = channel_count;
+
+    opus_decoder.Send(ADSP::Direction::DSP, ADSP::OpusDecoder::Message::InitializeDecodeObject);
+    auto msg = opus_decoder.Receive(ADSP::Direction::Host);
+    if (msg != ADSP::OpusDecoder::Message::InitializeDecodeObjectOK) {
+        LOG_ERROR(Service_Audio, "OpusDecoder returned invalid message. Expected {} got {}",
+                  ADSP::OpusDecoder::Message::InitializeDecodeObjectOK, msg);
+        return Service::Audio::ResultInvalidOpusDSPReturnCode;
+    }
+
+    auto error_code{static_cast<s32>(shared_memory.dsp_return_data[0])};
+    ASSERT(error_code <= OPUS_OK);
+    return ResultCodeFromLibOpusErrorCode[error_code + 7];

```

Comment:
Break this out as a helper function in the anonymous namespace and use a switch case for this:
```cpp
Result LibOpusErrorCodeToResult(u64 return_data) {
    switch (static_cast<s32>(return_data)) {
    case OPUS_OK:
        R_SUCCEED();
    case OPUS_BAD_ARG:
        R_RETURN(Service::Audio::ResultLibOpusBadArg);
    // ...
    }
}
```
then:
```suggestion
    R_RETURN(LibOpusErrorCodeToResult(shared_memory.dsp_return_data[0]));
```